mongodump/barrier.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (C) MongoDB, Inc. 2025-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+package mongodump
+
+import (
+	"os"
+	"time"
+
+	"github.com/mongodb/mongo-tools/common/log"
+	"github.com/pkg/errors"
+)
+
+// Wait until a file exists and can be opened for reading
+// This is used only for testing mongodump/mongorestore with resmoke
+// test infrastructure.  The tests will create the barrier file when they have
+// finshed writes to the source cluster.
+func waitForSourceWritesDoneBarrier(barrierName string) {
```

Comment:
This should accept a context, and whatever blocks in here should take the context into account. See my note below about making the `for` loop more idiomatic.

It should also probably return an error rather than panicking (see other note below).

Reply:
Nothing in mongodump is using any kind of Context.

Your suggestions are appropriate to designing a general-purpose barrier mechanism for wider use.  That is not the goal here: it is code which lives in the mongodump repo and is used in mongodump only for one specific tactical purpose, and is designed to do precisely that and with no complexity beyond that.

Reply:
We should design this particular piece to accept a context, then have callers pass in context.Background().

It is a defect of mongodump/restore that they generally don’t use Context. We shouldn’t propagate that to new code.

Reply:
That said, I suppose the only thing this function does that blocks (other than filesystem) is time.Sleep, which doesn’t take a context. (It surprises me that the stdlib still lacks a `time.SleepContext` or some other convenient way to do an interruptible wait.)